spring-cloud-sleuth-brave/src/main/java/org/springframework/cloud/sleuth/brave/bridge/BraveCurrentTraceContext.java

```diff
@@ -20,58 +20,70 @@
 import java.util.concurrent.Executor;
 import java.util.concurrent.ExecutorService;
 
+import brave.propagation.ThreadLocalCurrentTraceContext;
+
 import org.springframework.cloud.sleuth.CurrentTraceContext;
 import org.springframework.cloud.sleuth.TraceContext;
 
-/**
- * Brave implementation of a {@link CurrentTraceContext}.
- *
- * @author Marcin Grzejszczak
- * @since 3.0.0
- */
 public class BraveCurrentTraceContext implements CurrentTraceContext {
 
+	final ThreadLocal<Scope> scopes = new ThreadLocal<>();
+
 	final brave.propagation.CurrentTraceContext delegate;
 
 	public BraveCurrentTraceContext(brave.propagation.CurrentTraceContext delegate) {
 		this.delegate = delegate;
 	}
 
-	@Override
 	public TraceContext context() {
 		brave.propagation.TraceContext context = this.delegate.get();
+		return context == null ? null : new BraveTraceContext(context);
+	}
+
+	public CurrentTraceContext.Scope newScope(TraceContext context) {
 		if (context == null) {
-			return null;
+			clearScopes();
+			return Scope.NOOP;
 		}
-		return new BraveTraceContext(context);
+		CurrentTraceContext.Scope previous = this.scopes.get();
+		CurrentTraceContext.Scope current = new BraveScope(this.delegate.newScope(BraveTraceContext.toBrave(context)));
+		return new RevertingScope(this, previous, current);
```

Comment:
If `RevertingScope` deals with the `scopes` `ThreadLocal`, would not it be better to let it get the previous one on it's own instead of passing it? That way all logic that touched `scopes` would be in the same place.

Reply:
Can you elaborate? I don't think i follow

Reply:
Actually now i wonder if we shouldn't do it on the level of sleuth api rather than Braves. The same way we do it with micrometer tracing...

Reply:
I meant moving this:
```java
CurrentTraceContext.Scope previous = this.scopes.get();
```
Into the ctor of `RevertingScope` but maybe we can also move current:
```java
CurrentTraceContext.Scope current = new BraveScope(this.delegate.newScope(BraveTraceContext.toBrave(context)));
```

Reply:
We can't move all of that into reverting scpoe cause there's one reverting scope but two delegates (maybeScope, newScope). I can improve this constructor however. 

We also can't move the code to sleuth's api cause there are no implementations there - just interfaces. It has to stay here & in OTel.